functions/src/index.ts

```diff
@@ -0,0 +1,231 @@
+export type RoundDetails = {
+    url: string
+    tracks: GameTrack[]
+}
+
+export type GameTrack = {
+    id: string,
+    title: string,
+    artists: string[],
+    isAnswer: boolean,
+}
+
+export type GameDetails = {
+    name: string,
+    password: string,
+    hostId: string,
+    intermissionDuration: number,
+    roundDuration: number,
+    startTime: Date | null,
+    genre: string,
+    rounds: RoundDetails[],
+    leaderBoard: {
+        [id: string]: {
+            name: string,
+            score: number
+        }
+    }
+}
+
+export type SpotifyArtist = {
+    id: string,
+    name: string,
+    //The Spotify URI for the artist
+    uri: string
+}
+
+export type SpotifyTrack = {
+    id: string,
+    artists: SpotifyArtist[],
+    duration_ms: number,
+    explict: boolean,
+    name: string,
+    preview_url: string,
+    uri: string,
+}
+
+// Functions
+exports.createGame = functions.https.onRequest((request, response) => {
+    const gameGenre: string = String(request.query.genre)
+    const hostId: string = String(request.query.hostId)
+    const hostName: string = String(request.query.hostName)
+    const gameName: string = String(request.query.name)
+    const password: string = String(request.query.password)
+    const gameRounds: number = Number(functions.config().gameconfig.number_of_rounds)
+    const optionsPerRound: number = Number(functions.config().gameconfig.number_of_options)
+
+    getRoundDetails(gameGenre, gameRounds, optionsPerRound).then((roundDetails: RoundDetails[]) => {
+        const gameDetails: GameDetails = {
+            hostId: hostId,
+            name: gameName,
+            password: password,
+            genre: gameGenre,
+            rounds: roundDetails,
+            intermissionDuration: NaN,
+            roundDuration: NaN,
+            startTime: null,
+            leaderBoard: {
+                [hostId]: {
+                    name: hostName,
+                    score: 0
+                }
+            }
+        }
+        admin.firestore().collection("games").add(gameDetails).then(entry => {
+            response.send(entry.id)
+            return entry
+        }).catch(err => {
+            console.log(err)
+            throw err
+        })
+        return gameDetails
+    }).catch(error => {
+        console.log(error)
+        response.send(error)
+        return error
+    })
+})
+
+function getRoundDetails(gameGenre: string, gameRounds: number, optionsPerRound: number) {
+    return getSpotifyToken().then(res => {
+        const access_token = res
+        return getSpotifyTracks(gameGenre, gameRounds, access_token, optionsPerRound).then((tracks: SpotifyTrack[]) => {
+            const answerTracks: GameTrack[] = []
+            const answerUrls: string[] = []
+
+            // there is a chance that there will not be enough tracks with previews
+            // TODO: add fallback logic to get more tracks with previews
+            for (const track of tracks) {
```

Comment:
throughts on usnig this syntax tracks.forEach(track => { .... })?

Reply:
you can't break out of foreach loops, and it needs to break out of the loop once it finds tracks with previews equal to the number of rounds we have